repos/system_upgrade/el7toel8/libraries/tests/test_vsftpdutils.py

```diff
@@ -0,0 +1,82 @@
+import errno
+
+from leapp.libraries.common.vsftpdutils import get_config_contents, get_default_config_hash
+
+
+class MockFile(object):
+    def __init__(self, path, content=None, to_raise=None):
+        self.path = path
+        self.content = content
+        self.to_raise = to_raise
+        self.error = False
+
+    def read_file(self, path):
+        if path != self.path:
+            self.error = True
+            raise ValueError
+        if self.to_raise is None:
+            return self.content
+        else:
+            raise self.to_raise
+
+
+def test_getting_nonexistent_config_gives_None():
+    path = 'my_file'
+    f = MockFile(path, to_raise=IOError(errno.ENOENT))
+
+    res = get_config_contents(path, read_func=f.read_file)
+
+    assert not f.error
+    assert res is None
+
+
+def test_getting_inaccessible_config_gives_None():
+    path = 'my_file'
+    f = MockFile(path, to_raise=IOError(errno.EACCES))
```

Comment:
Nit: 
While having read_file function (leapp.libraries.common.vsftpdutils.read_file) implemented separately makes the read file functionality easily testable, having it passed as a callback into other methods like get_config_contents and get_default_config_hash imho is overcomplicating the api (honestly I can't think of a second way to read the file).
Have you considered using mock's context manager to control output of open() function instead of doing all the work yourself? 

Something like 
```
with mock.patch('builtins.open', side_effect=IOError) as open_mock:
    res = get_config_contents(path, read_func=f.read_file)
```

It's an open question, I might not have the full picture as I've definitely spent a lot less time with this patch than you had.

Reply:
Personally, I like the idea of injecting dependencies, which interact with the external system. It seems cleaner than making the test reliant on the way the implementation reads the files (another way to open the file would be `os.open`). But I do agree that my implementation of dependency injection is not ideal and makes it look a bit messy. I haven't found a better way so far.

Anyway, I did try mocking `builtins.open`, but couldn't get it to work. It seemed to me that apart from the function under test, something else was also using `open`, so mocking it broke that "something". But I may be wrong.

Reply:
@olysonek: `builtins.open` is kind of problematic. I thought it is just issue when you want to mock stuff and runnning whole actor in tests. But nowadays we have the `api` library, it is possible that there is hidden use of `open` as well in background.